DOCS 646: improve extend viam (#1029)
viamrobotics/docs: docs/program/extend/modular-resources/_index.md

---
title: "Create custom components and services as modular resources"
linkTitle: "Modular Resources"
image: "/tutorials/img/intermode/rover_outside.png"
imageAlt: "An intermode rover pictured outdoors."
weight: 10
type: "docs"
tags: ["server", "rdk", "extending viam", "modular resources", "components", "services"]
description: "Use the Viam module system to implement custom resources that can be included in any Viam-powered robot."
---

The Viam module system allows you to integrate custom [resources](/appendix/glossary/#term-resource) ([components](/components) and [services](/services)) into any robot running on Viam.

With modular resources, you can:

- Create new models of built-in component or service types
- Create brand new resource types

`viam-server` [manages](#modular-resource-management) modular resources configured on your robot like resources that are already built-in to the [Robot Development Kit (RDK)](/internals/rdk).

> Two key concepts exist across all Viam resources, built-in and modular, to facilitate this: [*APIs*](#apis) and [*models*](#models).

Reviewer: ```suggestion
Two key concepts exist across all Viam resources (both built-in and modular) to facilitate this: [*APIs*](#apis) and [*models*](#models).
```
Reads like those are the two concepts, so was a tad confusing.

## Key concepts

### APIs

> Every Viam [resource](/appendix/glossary/#term-resource) exposes an [Application Programming Interface (API)](https://www.ibm.com/topics/api).

Reviewer: [optional/opinion] IBM link is a weird place to go for a definition. Why not wikipedia or more obvious "reference" site?

Author: Ha true, will change to wiki

This can also be understood as the protocol that the resource "speaks".
Each API is described through <a href="https://developers.google.com/protocol-buffers" target="_blank">protocol buffers</a>.
Viam SDKs [expose these APIs](/internals/robot-to-robot-comms/).

Each Viam resource's API is uniquely namespaced as a colon-delimited-triplet in the form of `namespace:type:subtype`.

For example:

- The API of built-in component [camera](/components/camera) is `rdk:component:camera`, which exposes methods such as `GetImage()`.
- The API of built-in service [vision](/services/vision) is `rdk:service:vision`, which exposes methods such as `GetDetectionsFromCamera()`.

{{% alert title="Note" color="note" %}}

> You can see built-in Viam resource APIs in the <a href="https://github.com/viamrobotics/api" target="_blank">Viam GitHub</a>.

Reviewer 2: why is this not using mardown-style links?

Author: Just left over from og, will fix

[Viam SDKs](/program/sdk-as-client/) that expose these APIs are available in various programming languages and allow you to control your robots [securely from anywhere](/internals/robot-to-robot-comms/).
For example:

Viam APIs are uniquely namespaced, with each resource type represented as a *colon-delimited-triplet*.
For example, the built-in component type *camera*'s API is __rdk:component:camera__.
The built-in service type *vision*'s API is __rdk:service:vision__
- The API of built-in component [camera](/components/camera) is `rdk:component:camera`, which exposes methods such as `GetImage()`.
- The API of built-in service [vision](/services/vision) is `rdk:service:vision`, which exposes methods such as `GetDetectionsFromCamera()`.

{{% alert title="Note" color="note" %}}
You can see built-in Viam resource APIs in the <a href="https://github.com/viamrobotics/api" target="_blank">Viam GitHub</a>.
{{% /alert %}}

### Models

A model is an implementation of a resource type that implements all or some of the API methods of a resource type API.
Models allow you to control any number of versions of a given resource with a consistent interface.

> This is powerful, because you only need to learn and code against one interface which remains the same for different models of the same component type.

Reviewer 2: That last sentence is important to call out somewhere

Author: It is called out on line 46! `Models allow you to control different versions of resource types with a consistent interface.`
and the examples following

> A *model* descrines a specific implementation of a resource that implements this resource's API.

Reviewer: ```suggestion
A *model* describes a specific implementation of a resource that implements (speaks) an API.
```

> - Regardless, any motor model that implements the *rdk:component:motor* API can be powered with the `SetPower()` method.

Reviewer 2: This is one exaple, so this should not be in bullet points

Add a modular resource to your robot configuration in five steps:

> 1. Code a module in Go or Python, using the module support libraries provided by the Python or Go [Viam SDK](/program/sdk-as-client).

Reviewer 2: This feels a bit draw the owl 
![image](https://user-images.githubusercontent.com/5212232/235166601-d23d643d-1027-46ad-a4dd-177721c8d98d.png)

We should probably point to a template or something - can you make a follow up ticket for that?

Reviewer 2: Oh it's covered below - can you put a link on code a module to link to the next section

> 1. Code a module in Go or Python that implements a new resource and registers the component in the Viam RDK's [global registry of robotic parts](https://github.com/viamrobotics/rdk/blob/main/registry/registry.go).

Reviewer: We do not YET have a "global registry" which is a concept in the planning stage where people will be able to upload/download/share modules with others.

The registry concept linked here is specific to Go, so isn't a good link to provide unless detailing out go specific implementation.

I would suggest just cutting this short to "Code a module in Go or Python, using the module support libraries provided by our SDKs." and then perhaps link to the relevant module docs/sections for the two SDKs.

Author: Ok perfect thanks

> 2. Create an executable file that runs your module.

Reviewer: This sounds the same as above when worded this way. Perhaps "Compile or package the module code into an executable."

Author: Sounds good done

3. Save the executable in a location your `viam-server` instance can access.

> 4. Add a **module** referencing this executable to the configuration of your robot.

Reviewer 2: in the app, right? Should probably specify that

Reviewer 2: actually all of these should just also have a link to the section where there's more info

5. Add a new component or service referencing the custom resource provided by the configured **module** to the configuration of your robot.

{{% alert title="Modules vs. modular resources" color="tip" %}}

A configured **module** can make one or more *modular resources* available for configuration.

{{% /alert %}}

### Code your module

Code a module in the Go or Python programming languages with [Viam's SDKs](/program/sdk-as-client) that does the following:

> 1. Implements a new resource, including implementing any methods the Viam RDK defines for the API of a built-in type if defining a new model (ex. `rdk:component:base`)

Reviewer: I would expand this to multiple steps that actually explain what's needed. Otherwise it feels like "How to draw an owl" ( https://seths.blog/2014/01/how-to-draw-an-owl/ )

Tentative breakdown might look like:

1. If creating a new API, define its messages and methods in protobuf, then generate code in your chosen language (Go or Python) and use the generated code to implement the higher level server and client functions required.

2. Create a new resource model, making sure to implement all required methods defined in the chosen API (either your own, or a pre-existing API.)

3. Create a main program that will serve as the module itself, using the module helpers provided by your chosen SDK

4. Import both the required API(s) and model(s) into the main program, and register them with the module helper SDK.

5. Compile and/or package your program.

> 2. Validates the module and registers the component in the Viam RDK's [global registry of robotic parts](https://github.com/viamrobotics/rdk/blob/main/registry/registry.go).

Reviewer: Again, we have no "global registry" and the link here is to a go-specific registry that's a different concept (and also go specific.) I also don't understand what is meant by "validate the module" here.

Author: Ahhh I meant the validate function and was mistaken about the registry info! So this is great feedback, will change

    # is_moving: checks if either motor on the base is moving with motors' is_powered
    async def is_moving(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> bool:
        return self.left.is_powered(extra=extra, timeout=timeout)[0] or self.right.is_moving(extra=extra, timeout=timeout)[0]
```

<file>init.py</file>

``` python {class="line-numbers linkable-line-numbers"}
"""
This file registers the MyBase model with the Viam Registry.
"""

from viam.components.motor import *
from viam.components.base import Base
from viam.resource.registry import Registry, ResourceCreatorRegistration

from .my_base import MyBase

Registry.register_resource_creator(Base.SUBTYPE, MyBase.MODEL, ResourceCreatorRegistration(MyBase.new, MyBase.validate_config))
```

{{% /tab %}}
{{< /tabs >}}

{{% /expand%}}

{{% alert title="Caution" color="caution" %}}

You must define all functions belonging to a built-in resource type if defining a new model. Otherwise, the class won’t instantiate.

- If you are using the Python SDK, raise an `NotImplementedError()` in the body of functions you do not want to implement or put `pass`.

> - If you are using the Go SDK, return `errUnimplemented` or leave the body of functions you do not want to implement empty.

Reviewer: Go function bodies can't be empty. And functions that have a return signature must follow that signature. MOST can probably return just errUnimplemented, but others may have to create and return correctly typed (but empty) values.

Author: Ok addressed, thanks!

{{% /alert %}}

### Make your module executable

To add a module to your robot, you need to have an [executable file](https://en.wikipedia.org/wiki/Executable) that runs your module when executed, can take a local socket as a command line argument, and cleanly exits when sent a termination signal.
Your options for completing this step are flexible, as this file does not need to be in a raw binary format.

One option is creating and save a new shell script (<file>.sh</file>) that runs your module.
For example:

{{< tabs name="Template Shell Scripts as Module Executables" >}}
{{% tab name="Go" %}}

``` shell
#!/bin/sh
cd <path-to-your-module-directory>

go build ./
# Be sure to use `exec` so that termination signals reach the go process,
# or handle forwarding termination signals manually
exec ./<your-module-directory-name> $@
```

> {{% /tab %}}

Reviewer: This is not a great example for go. Go actually builds binaries that can be directly run. So when compiling/building the module, they'll get an actual binary. No need to try to compile it here in a shell script and then exec it.

(If you got this from RDK examples, that's a tricky workaround I set up testing, not something you'd want to do during deployment, so my apologies.)

Author: Ok! Addressed-- lmk if terminology is still off. Not great at go but learning

{{% tab name="Python" %}}

``` shell
#!/bin/sh
cd <path-to-your-module-directory>

# Be sure to use `exec` so that termination signals reach the python process,
# or handle forwarding termination signals manually
exec python3 <your-module-directory-name>.main $@
```

{{% /tab %}}
{{< /tabs >}}

To make this file executable, run the following command in your terminal:

``` shell
sudo chmod +x <FILEPATH>/<FILENAME>
```

### Make sure `viam-server` can access your executable

Ensure that the code defining your module is saved where the instance of `viam-server` behind your robot can read and execute it.

For example, if you are running `viam-server` on an Raspberry Pi, you'll need to save the module on the Pi's filesystem.

Obtain the path to the executable file on your computer's filesystem by running the following commands in your terminal:

``` shell
cd <path-to-your-module-directory>

> pwd

Reviewer: This gives you the directory where your module is, not the full path (that goes in the config) which includes the directory AND the name of the module file.

Author: Ok addressed I think!

Reviewer: No, I think there was a misunderstanding. 

What I originally meant is that this shell script is not doing what you say it is. The text says it is how you "obtain the path to the executable file", but it's not doing that. It is, in fact, only obtaining the part of the path; the directory the executable is in. You still must append the executable name. You can't just cd to the executable itself either (per this change), since it's not a directory.

You probably just want to instruct people to "get the full path to your executable" and let them know that if they see it in their terminal, they can use realpath. E.g. `realpath myModule` or `realpath moduleDirectory/myModule`

Author: Oh yes I did misunderstand this/wasn't paying attention to the rest of the instructions here-- thanks for the suggestion/fixed

> Obtain the path to the executable file on your computer's filesystem by running the following commands in your terminal:

Reviewer: ```suggestion
Obtain the real (absolute) path to the executable file on your robot/pi filesystem by running the following command in your terminal:
```
Minor suggestion for some clarification.
